gc/2021/GC-05-18.md

```diff
@@ -0,0 +1,186 @@
+
+AR: ocaml, typescript, Scala but it uses JVM reflection
+
+BT: order dependence, doesn't it mean if you export a type, you need to define in minimal recursion groups
+
+AR: You don’t have to, but you probably want to. You could define all the types in your module in one big recursion group, but that’s less useful. 
+
+JK: on order depdennce, in first design it is clear we need it, with flatten index is it still order dependence? why?
+
+AR: Yes because that dependence is ingrained in this structure that is a tuple over types. Comparing equality if the tuple is not ordered would be complicated.
+
+RT: any case studies done with this?
+
+AR: no
+
+RT: looked this up in OO languages?
+
+AR: nominally type OO languages, no problem. It is as expressive as nominal plus more, any nominal can be mapped. You need this at module boundaries for linking for nominal types, essentially iso recursive at link type, that's what you check against.
+
+RT: What about subtyping? How do you deal with a non-recursively defined class being extended by a class that is mutually recursive with other classes.
+
+	class A
+	rec-class B <: A
+	B and C are mutually recursive
+
+AR: what would be the problem? it wouldn't get recursive just by extending it. B and C would be recursive but not with A.
+
+RT: Mutually recursive group, a member of that, a non thing, and they are not lining up. How is B a subtype of A?
+
+BT: subtyping is totally orthogonal, recursive relation over types, only time it comes up is immutable field, anything else is a check. The recursion we have now for width and depth subtyping for structs, just works out.
+
+TL: no problem having a type in a recursion group, be a subtype of another type not in a recursion group?
+
+AR: this might be actually a reason to allow nesting of recursion groups. Because you basically have recursion over recursive things. The subtyping rule for iso-rec is one rec group is subtype of another rec-group if you have both depth and width subtyping on components. Means that you can only extend one group in one dimension at a time. But if you have two mutually recursive classes, then the recursion group, for each clas the recursion between methods, then recursion between the two classe, which is one level up. The inner one want to be a subtype, but not on the combined thing.
+
+RT: I understand it's complicated, asking if you have worked through to see if it works
+
+AR: no, not in that detail. I'm pretty confident that it works sufficiently well, iso-recursive is often used as the type theorteic model of nominal types
+
+RT: some red flags in representation. E.g. non-uniform recursion and how it is decidable. But it is known that non-uniform recursion that can handle OO subtyping, then it is undecidable
+
+CW: <something> bounded quantification
+
+RT: No, generics are undecidable with iso-recursive types without bounded quantification. Paper by <missed> names I presented to the group back in January.
```

Comment:
@RossTate, can you fill in the paper author here?

Reply:
```suggestion
RT: No, generics are undecidable with iso-recursive types without bounded quantification. Follows from the [paper](https://kar.kent.ac.uk/58183/7/javats.pdf) by Radu Grigore as I presented to the group back in [December](https://github.com/WebAssembly/meetings/blob/main/gc/2020/GC-12-15.md).
```
Here you go! Though, looking back, apparently it was December and not January.

Reply:
Excellent, thanks!